text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+  It runs tests in specially prepared environments.
+  To script the setup and the test execution, it uses a custom subshell wrapper to execute CLI commands.
+
+#### C. Jupyter Notebooks
+
+Writing a Python Jupyter Notebook that uses complex computations on large datasets.
+These computations need to happen in the cloud using EMR.
+The EMR cluster and tasks can be defined and deployed from within the Notebook.
+API calls are then used to execute the tasks, download results and remove the infrastructure.
+
+Today users are limited to make imperative SDK calls or to provide a separate companion app with additional instructions.
+With the [Programmatic Toolkit], CDK apps can be deployed directly from within a Notebook and resources can be referenced directly as the result of CDK actions.
+
+Real world examples:
+
+* [Boto3 with Jupyter Notebook](https://blog.jimmyray.io/boto3-with-jupyter-notebook-721a01eab285)
+* [Using Jupyter Notebooks as your cloud IDE for DevOps](https://guyernest.medium.com/using-jupyter-notebooks-as-your-cloud-ide-for-devops-a49c03e5cb3f)
+
+## Internal FAQ
+
+### Why are we doing this?
+
+The [Programmatic Toolkit] unlocks new scenarios for programmatic use of the AWS CDK.
+Today integrators can call the AWS CDK CLI via a shell or use `cli-lib-alpha` for programmatic access.
+Both approaches limit integrators to inherit the exact behavior of the CLI, including all interaction patterns with the user.
+This is an obvious mismatch: Programmatic integrations don't require interaction, unless that's the specific use case an integrator is building.
+Additionally the interface `cli-lib-alpha` offers is not optimized for programmatic use, further limiting its usefulness.
+
+The [Programmatic Toolkit] aims to solve these limitations by providing a proper programmatic interface to CDK's functionality,
+sallowing integrators to have full control over all output and combine CDK actions with custom steps in their applications.
+With this project, we are changing the paradigm.
+Programmatic usage will become an equally considered use case, in fact the proposed implementation will require it to be considered first.
+
+### Why should we _not_ do this?
+
+There are requirements that cannot be satisfied with the current available options.
+Specifically event hooks, control over logging, structured errors and typed returns are not possible (refer to the appendix for details).
+
+Implementing the [Programmatic Toolkit] proposal will make it more difficult to add new features to the CLI.
+It creates a new level of abstraction, that needs to be considered going forward.
+The complexity of these abstractions might discourage community contributions and increase the time to build new features.
+
+We are only aware of a small number of interested integrators,
```

Comment:
This statement resonates with me, and raises some concerns. To alleviate them, we should present alternative solutions to the new specific requirements we have.

Reply:
We have specific requirements. Alternatives were considered. This sections is why we shouldn't do this and the answer here is implying that the technical complexity may not outweigh the benefits. However I'm clearly proposing we should do this and in order to satisfy requirements we have to do something, and I think it should be this proposal.

Maybe I'm missing your point though.